MoreLinq/IntersectAll.cs

```diff
@@ -0,0 +1,182 @@
+        }
+
+        /// <summary>
+        /// Returns the sequence of elements from the first sequence
+        /// whose keys are in the second sequence, according to the given key selector.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the source and result elements.</typeparam>
+        /// <typeparam name="TKey">The type of the key.</typeparam>
+        /// <param name="first">The sequence of potentially included elements.</param>
+        /// <param name="second">The sequence of keys which may allow elements in
+        /// <paramref name="first"/> to be returned.</param>
+        /// <param name="keySelector">The mapping from source element to key.</param>
+        /// <returns>Sequence of elements from first sequence
+        /// whose keys are in second sequence, according to the given key selector.</returns>
+        /// <remarks>
+        /// This is a sequence operation; if multiple elements in <paramref name="first"/> have
+        /// equal keys, all such elements are returned.
+        /// This operator uses deferred execution and streams the results, although
+        /// a set of keys from <paramref name="second"/> is immediately selected and retained.
+        /// </remarks>
+        public static IEnumerable<TSource> IntersectAllKeys<TSource, TKey>(this IEnumerable<TSource> first,
+            IEnumerable<TKey> second,
+            Func<TSource, TKey> keySelector) {
+
+            if (first == null) throw new ArgumentNullException("first");
+            if (second == null) throw new ArgumentNullException("second");
+            if (keySelector == null) throw new ArgumentNullException("keySelector");
+
+            return IntersectAllKeysImpl(first, second, keySelector, null);
+        }
+
+        /// <summary>
+        /// Returns the sequence of elements from the first sequence
+        /// whose keys are in the second sequence, according to the given key selector.
+        /// </summary>
+        /// <typeparam name="TSource">The type of the source and result elements.</typeparam>
+        /// <typeparam name="TKey">The type of the key.</typeparam>
+        /// <param name="first">The sequence of potentially included elements.</param>
+        /// <param name="second">The sequence of keys which may allow elements in
+        /// <paramref name="first"/> to be returned.</param>
+        /// <param name="keySelector">The mapping from source element to key.</param>
+        /// <param name="keyComparer">The key comparer. If <c>null</c>, uses the default TKey equality comparer.</param>
+        /// <returns>Sequence of elements from first sequence
+        /// whose keys are in second sequence, according to the given key selector.</returns>
+        /// <remarks>This is a sequence operation; if multiple elements in <paramref name="first"/> have
+        /// equal keys, all such elements are returned.
```

Comment:
What about `second` duplicates? I think it should be mentioned that duplicates there are discarded.